### PR TITLE
Feat: Migration Paraswap V5

### PR DESCRIPTION
**Pull request by n4l5u0r**
- Addition of Paraswap V5 Smart Contract additional methods.
- Adding some test cases for Paraswap V4 Smart Contract methods.
- Test code refactoring
- Sonar identified issues fix.

### CODE REVIEW

File: src/handle_query_contract_ui.c [hunk 1]
```diff
@@ -6,8 +6,8 @@ static void prepend_ticker(char *dest, size_t destsize, const char *ticker) {
     if (dest == NULL || ticker == NULL) {
         THROW(0x6503);
     }
-    uint8_t ticker_len = strlen(ticker);
-    uint8_t dest_len = strlen(dest);
+    size_t ticker_len = strnlen(ticker, MAX_TICKER_LEN);
+    size_t dest_len = strnlen(dest, destsize - ticker_len);
```
Comment (jibeee) [inline src/handle_query_contract_ui.c:10]: I am not sure you should use `destsize - tickerlen`. Doing this, the condition just below will always be true, and the resulting amount will be truncated.

Correct code is:
```c
    size_t dest_len = strnlen(dest, destsize);
```

**Review by jibeee — CHANGES_REQUESTED** [review]
I am not sure you should pass `destsize - tickerlen` as a size to `strnlen`. Doing this, the condition just below will always be true, and the resulting amount will be truncated.

File: src/handle_query_contract_ui.c (hunk 1 above)
Comment (n4l5u0r) [reply src/handle_query_contract_ui.c]: fix applied

**Review by jibeee — APPROVED** [review]

**Comment by pscott** [comment]
I believe you could remove the function `prepend_ticker` and simply replace it with `amountToString`, [just like so](https://github.com/LedgerHQ/app-plugin-boilerplate/blob/2d1bab2ebc81d33e20f0674de64f2cd1484257b5/src/handle_query_contract_ui.c#L8-L17).

**Comment by alacombe-ledger** [comment]
@n4l5u0r would you be able to solve the conflicts and take a look at pscott suggestion? thank you :pray:

**Comment by n4l5u0r** [comment]
Yes no problem let us check that

**Comment by n4l5u0r** [comment]
@pscott which version of app-ethereum & ethereum-plugin-sdk should we consider for stable integration tests ?

**Comment by n4l5u0r** [comment]
> @pscott which version of app-ethereum & ethereum-plugin-sdk should we consider for stable integration tests ?

@pscott @alacombe-ledger 
Current stability of app-ethereum on Nano-X compilation blocks nano-x plugin build : https://github.com/LedgerHQ/app-ethereum/runs/4037174687?check_suite_focus=true

**Comment by alacombe-ledger** [comment]
@TamtamHero could you please advise what we can do here? thank you!

**Comment by n4l5u0r** [comment]
@alacombe-ledger we solved the conflicts but had to comment out for the moment the tests for Nano-X as we don't have yet a stable version of app-ethereum using BOLOS_SDK=$NANOX_SDK
Once we reach the stability point we will proceed with updates on the Nano-X part.

**Comment by pscott** [comment]
@n4l5u0r hi and thank you for your work :)
Indeed the nanox-secure-sdk isn't working properly with the CI right now. However, since your code does not introduce LNX specific logic, it's safe to say that it should work without any concerns :)

Regarding my previous comment: I think you could remove `prepend_ticker` and use `amountToString` juste like so 
https://github.com/LedgerHQ/app-plugin-boilerplate/blob/ec1cc4ea621402c554ba585ef95206200d91e6d5/src/handle_query_contract_ui.c#L24-L38

Thanks!

**Comment by n4l5u0r** [comment]
Hi @pscott 

> Regarding my previous comment: I think you could remove `prepend_ticker` and use `amountToString` juste like so
> https://github.com/LedgerHQ/app-plugin-boilerplate/blob/ec1cc4ea621402c554ba585ef95206200d91e6d5/src/handle_query_contract_ui.c#L24-L38

-  The `amountToString` is used already in `handle_provide_parameter.c` so the change will impact a little bit more the overall code.

> EDIT: you can also remove the `strlcpy` in `handle_provide_token.c` since you will use the network ticker directly in `handle_query_contract_ui` :D

- the use of network ticker remains conditional need to handle `context->ticker` set if `ADDRESS_IS_NETWORK_TOKEN` is not met

**Comment by pscott** [comment]
> Hi @pscott
> 
> > Regarding my previous comment: I think you could remove `prepend_ticker` and use `amountToString` juste like so
> > https://github.com/LedgerHQ/app-plugin-boilerplate/blob/ec1cc4ea621402c554ba585ef95206200d91e6d5/src/handle_query_contract_ui.c#L24-L38
> 
> * The `amountToString` is used already in `handle_provide_parameter.c` so the change will impact a little bit more the overall code.
> 
> > EDIT: you can also remove the `strlcpy` in `handle_provide_token.c` since you will use the network ticker directly in `handle_query_contract_ui` :D
> 
> * the use of network ticker remains conditional need to handle `context->ticker` set if `ADDRESS_IS_NETWORK_TOKEN` is not met

Indeed it will affect the code but it will reduce the app size and make the code cleaner :)
the `strlcpy` I was referring to in the second part of my answer was the one which copies `DEFAULT_TICKER`. I think you can still remove this one, and in `handle_query_contract_ui` adapt the code :D

The thing with `DEFAULT_TICKER` is that it will always be `ETH ` even if you're running a tx on the binance chain or polygon :)

**Comment by alacombe-ledger** [comment]
@n4l5u0r fyi the issue with nanox-secure-sdk has been fixed

**Comment by n4l5u0r** [comment]
> @n4l5u0r fyi the issue with nanox-secure-sdk has been fixed

Ok let us check then the test suite upgrade with this fix before proceeding with the merge

**Comment by n4l5u0r** [comment]
> > @n4l5u0r fyi the issue with nanox-secure-sdk has been fixed
> 
> Ok let us check then the test suite upgrade with this fix before proceeding with the merge

Hi @pscott @alacombe-ledger 

looks like Zemu 16.5 does not include the support for 2.0 for NanoX yet: https://github.com/LedgerHQ/app-ethereum/runs/4192950082?check_suite_focus=true

Do you have somewhere any working repo with the last zemu 20.2 support ?

In the meantime we are refactoring the code mentioned previously : https://github.com/LedgerHQ/app-plugin-paraswap/pull/4#issuecomment-968292794

**Comment by n4l5u0r** [comment]
> > Hi @pscott
> > > Regarding my previous comment: I think you could remove `prepend_ticker` and use `amountToString` juste like so
> > > https://github.com/LedgerHQ/app-plugin-boilerplate/blob/ec1cc4ea621402c554ba585ef95206200d91e6d5/src/handle_query_contract_ui.c#L24-L38
> > 
> > 
> > 
> > * The `amountToString` is used already in `handle_provide_parameter.c` so the change will impact a little bit more the overall code.
> > 
> > > EDIT: you can also remove the `strlcpy` in `handle_provide_token.c` since you will use the network ticker directly in `handle_query_contract_ui` :D
> > 
> > 
> > 
> > * the use of network ticker remains conditional need to handle `context->ticker` set if `ADDRESS_IS_NETWORK_TOKEN` is not met
> 
> Indeed it will affect the code but it will reduce the app size and make the code cleaner :) the `strlcpy` I was referring to in the second part of my answer was the one which copies `DEFAULT_TICKER`. I think you can still remove this one, and in `handle_query_contract_ui` adapt the code :D
> 
> The thing with `DEFAULT_TICKER` is that it will always be `ETH ` even if you're running a tx on the binance chain or polygon :)

Hi @pscott, @alacombe-ledger,

The requested code refactoring has been completed.
We can keep the PR open until the Zemu 16.5 & 2.0 SDK misalignment is resolved and we could then validate Nano X test suite.

**Comment by TamtamHero** [comment]
Hello !
Could you please give me edit rights on this PR @n4l5u0r ?

**Comment by n4l5u0r** [comment]
> Hello ! Could you please give me edit rights on this PR @n4l5u0r ?

Hello @TamtamHero, as this is an organization owned fork I added you as contributor to the fork.
Let me know if you have sufficient access.

**Comment by TamtamHero** [comment]
Nope, still can't push
But you should be able to give me these rights directly from this PR page (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

**Comment by n4l5u0r** [comment]
> Nope, still can't push But you should be able to give me these rights directly from this PR page (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

@TamtamHero this works for user owned forks and there is a known limitation for organization owned forks (https://github.com/isaacs/github/issues/1681). I increased your level of authorization on the fork repo to `Maintain`, you should have received an invite to accept. With this level you could be able to push directly to master which is not protected on the fork repo.

**Review by jibeee — APPROVED** [review]

**Comment by alacombe-ledger** [comment]
@n4l5u0r can you please quickly confirm that v4 is not blind signing on your side? Thank you

**Comment by n4l5u0r** [comment]
> @n4l5u0r can you please quickly confirm that v4 is not blind signing on your side? Thank you
Not blind signing on our side. You can run tests on v4 only in the tests/src/v4 folder to see the output.

**Review by apaillier-ledger — APPROVED** [review]

**Pull request merged**